m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -555,14 +561,18 @@ private void flushBuffered() {
                                 .setCommonTags(commonTags)
                                 .setMetrics(metricsBuffer)
                 );
-            } catch (TException tException) {
-                LOG.warn("Failed to flush metrics: " + tException.getMessage());
+            } catch (Throwable t) {
```

Comment:
Is the TException -> Throwable change intentional? If so, what motivated it?

Reply:
To prevent uncaughts in the processor thread

Reply:
were there particular uncaughts that triggered this? wondering why it's required now?

Reply:
B/c failing processor thread due to `RuntimeException` doesn't make sense

Reply:
Hm I don't agree with that. E.g. why should this catch an `OutOfMemoryError` ?

Reply:
If you don't agree, would be great if you provide the use-case and explain the rationale.

Even with `OutOfMemoryError` i don't see the reason of having fly-away exception that crashes the reporter's thread.

Reply:
The purpose of a try catch is to recover from failure. How are you recovering from an OOM? 

More generally, catching throwable is one of those things people strongly recommend. E.g. https://stackoverflow.com/questions/6083248/is-it-a-bad-practice-to-catch-throwable 

It would be great if you could explain your rationale for want to do this.

Reply:
Sorry, but i struggle to understand where the disconnect is: i had already explained the rationale -- **to make sure that reporter thread isn't crashed b/c of uncaught exception**

Reply:
Same here, I don't understand why we seem to be talking over each other. 

`Throwable` is the super class for all exceptions in the JVM. Catching it catches every mode of failure the JVM permits. This includes catching things you shouldn't, one e.g. of this is an OOM. If the JVM is throwing an OOM, the process is having much bigger issues than a metrics library malfunctioning. As a user, I want/expect the process to crash, not trudge along in a degraded state. An OOM isn't the only such error (e.g. if this gets a NPE, there's something messed up with the library, and we want it to crash loudly).

I'm not against making the exception classes being caught wider, but I am against catching unrecoverable errors. 

Is this changed motivated by real world failures? If so, what have you observed?

Reply:
Let me start off to say, that i'm not entirely sure where you noticed occasions of talking over each other. I appreciate your in-depth deliberation, but at the same time would like to point out that most of the questions you raise could be easily addressed on your own.

> This includes catching things you shouldn't, one e.g. of this is an OOM. If the JVM is throwing an OOM, the process is having much bigger issues than a metrics library malfunctioning. As a user, I want/expect the process to crash, not trudge along in a degraded state.

Agree on the premise that loud failure is preferred, however your conclusion is wrong -- uncaught exception in the thread will not crash the process, it will only crash the reporting thread meaning that application will be left without metrics.

Your example w/ OOM is invalid in nature: library shouldn't be trying to handle the issues it couldn't handle, this should be occupation of the application itself. 

Re-iterating on the rationale for this change: producer's thread should be resilient to exceptions, and continue attempting to produce metrics even if each attempt has low chance of success as this is the only way for application to obtain telemetry.

Reply:
re: "would like to point out that most of the questions you raise could be easily addressed on your own."
Please call out specific examples. 

re: talking over each other -- bad phrasing. I was expressing that I didn't understand our disconnect. I do now based on your point about the JVM crashing the Thread, not the Process when the failure occurs. Good callout.

- Is this practice (logging such errors and continuing) idiomatic in Java?
- Does the thread always end up in a state it can continue executing? i.e. should it create a replacement?

Reply:
1. This practice has nothing to do with idioms of Java. This approach allows to keep the the processor running even in the face of some transient/permanent issues.
2. I don't really understand your question. Why would it not be able to continue?

Reply:
1. Idioms/prior art are a good indication of what standard practice is in the community. I don't like libraries doing things against the norm. 
2. You're catching literally the lowest level of error the JVM offers, it could catch any manner of things corrupting thread state -- e.g. StackOverflowError would be caught here, is the thread able to continue execution after catching that? Is that true for ALL errors? 

All in all, I haven't seen a strong case that this helps in. It's coming down to a style preference: yours to continue trying, and mine is to log and stop. Until I see a real world case where this helps in, I'm against the making it the default behaviour, further, it's better from a b/w compatibility standpoint as well. That said, I'm not against the library supporting if you want it. 

So if you want to include this, please make the library offer a choice so that it's injectable by the user, and override in your service.

Reply:
Spoke offline, a good case was brought up that such resilience could actually obscure the problems within the library as, potentially leaving it in a slightly degraded state that might be hard to trace back.